pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,363 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+import (
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/kubestellar/kubestellar/api/edge/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+const clusterScopedNamespace = ""
+
+// placementResolution stores the selected objects and destinations for a single placement.
+// The mutex should be read locked before reading, and write locked before writing to any field.
+type placementResolution struct {
+	sync.RWMutex
+
+	// map key is GVK/namespace/name as outputted by util.key.GvkNamespacedNameKey()
+	objectIdentifierToKeyMap map[string]*util.Key
```

Comment:
(A) What goes wrong if you drop "Map" from this field's name?

(B) The point of this field is to hold a set of Key, right? Why not use `Set[Key]` (importing https://github.com/kubernetes/apimachinery/blob/v0.28.2/pkg/util/sets/set.go#L24).

Reply:
(A) is fine by me but I dont think this is critical.

(B) unless I explicitly implement Key comparison, the field DeletedObject would be problematic. Therefore I wonder if it's a necessary change.

Reply:
I see that you made change (A). Regarding (B), I do not follow your remark about implementing Key comparison. The go language does not allow user-defined comparison and hashing behavior for its built-in `map`s (I sure wish it did!).  As long as a a Key is not a pure object reference, the way to avoid string hackery would be to ensure that every domain value in the map has no extraneous content. One approach would be to define `func clean(Key) Key` that removes the DeletedObject if present, and apply `clean` before indexing the map. Another approach would be to factor Key into a pure (clean) part plus the Deleted object, and use only the pure part for indexing the map.  For example:

```go
type ObjectRef struct {
	GVK            schema.GroupVersionKind
	NamespacedName cache.ObjectName
}
type Key struct {
	ObjectRef
	DeletedObject  *runtime.Object
}
```

Reply:
If I understood correctly, in B you suggest I switch to Set[Key] instead of map[string]*Key, right? To use Key in Set, it would need to implement Comparable, which it does, since all its fields are Comparable. BUT, the problem is two Keys can refer to the same object, with DeletedObject changing, thus making them different keys.

The latter is not a problem in the current implementation because I use an object-unique string to map Keys.

I agree that Key must be reworked, and perhaps a lot of the "buggy" flow can be discarded when we delete the transport code from the placement controller. But I think it should be in the followup refactoring work I referred to in a previous comment. What do you think?

Reply:
I wrote two ways that you could have a map indexed by values where the `DeletedObject` field is not an issue, while still having Paolo's logic and transport code.

This is an FYI, does not have to be done in this PR. Can indeed be done after removing the transport code.